File: MIDIJava/src/io/github/seanboyy/Midi/TrackChunk.java
package io.github.seanboyy.Midi;

import java.util.ArrayList;

public class TrackChunk implements Chunk {

    private int length;

    private ArrayList<Event> events;


    ArrayList<Event> getEvents() {
        return events;
    }

    TrackChunk(int length, ArrayList<Event> events) {
        this.length = length;
        this.events = new ArrayList<>();
        this.events.addAll(events);
    }

    @Override
    public byte[] toBitString() {
        byte[] bitString = new byte[getLength()];
        int[] cursor = {0};
        bitString = StreamingHelper.streamValueToBitString(cursor, bitString, MidiConstants.TRACK_SIGNATURE);
        bitString = StreamingHelper.streamValueToBitString(cursor, bitString, length);
        for (Event e : events) {
            bitString = StreamingHelper.streamObjectToBitString(cursor, bitString, e, e.getLength());
        }
        return bitString;
    }

    @Override
    public int getLength() {
        return 8 + length;
    }
}

File: MIDIC#/MIDILibrary/MIDILibrary/TrackChunk.cs
using System.Collections.Generic;

namespace MIDILibrary
{
    public class TrackChunk : IChunk
    {
        private readonly uint _length;
        public List<IEvent> Events { get; }

        public TrackChunk(uint length, IEnumerable<IEvent> events)
        {
            _length = length;
            Events = new List<IEvent>();
            Events.AddRange(events);
        }
        
        public uint GetLength()
        {
            return 8 + _length;
        }

        public byte[] ToBitString()
        {
            byte[] bitString = new byte[GetLength()];
            uint cursor = 0;
            StreamingHelper.StreamValueToBitString(ref cursor, ref bitString, MidiConstants.TRACK_SIGNATURE);
            StreamingHelper.StreamValueToBitString(ref cursor, ref bitString, _length);
            foreach (var e in Events)
            {
                StreamingHelper.StreamObjectToBitString(ref cursor, ref bitString, e, e.GetLength());
            }
            return bitString;
        }
    }
}
